### ISSUE TITLE
sql: implicit type cast during comparison

### ISSUE THREAD

**Issue by Korablev77** [tarantool/tarantool#4230]
We've revised current state of implicit conversions, and found that several rules are seem to be
exceptional from our rules. Here's a list of what should be fixed:
- [ ] BLOB to INT and BLOB to REAL conversions should be disallowed. Now comparisons between these types lead to false result:
```
CREATE TABLE t (s1 INT PRIMARY KEY, s2 INTEGER);
INSERT INTO t VALUES (-1,NULL),(0,0);
SELECT * FROM t WHERE s2 = X'';
---
- metadata:
  - name: S1
    type: integer
  - name: S2
    type: integer
  rows: []
...
```

- [ ] TEXT to INT and TEXT to REAL conversions should be allowed only if value of string can be converted to numeric literal. Otherwise, error should be raised. Now it is always legal:

```
SELECT * FROM t WHERE s2 = '';
---
- metadata:
  - name: S1
    type: integer
  - name: S2
    type: integer
  rows: []
...
```

- [ ] Conversion from any type to SCALAR should be allowed. Now it depends on content held in table:

```
CREATE TABLE t (s1 INT PRIMARY KEY, s2 SCALAR);
INSERT INTO t VALUES (-1,NULL),(0,0),(1,FALSE),(2,1.1),(3,X''),(4,''),(5,'');
tarantool> SELECT * FROM t WHERE s2 = TRUE;
---
- error: 'Type mismatch: can not convert INTEGER to boolean'
...

tarantool> SELECT * FROM t WHERE s2 = 1;
---
- error: 'Type mismatch: can not convert INTEGER to boolean'
...

tarantool> SELECT * FROM t WHERE s2 = 1.1;
---
- error: 'Type mismatch: can not convert REAL to boolean'
...

tarantool> SELECT * FROM t WHERE s2 = X'';
---
- error: 'Type mismatch: can not convert BLOB to boolean'
...

tarantool> SELECT * FROM t WHERE s2 = '';
---
- error: 'Type mismatch: can not convert TEXT to boolean'
...
```
Last point can be done after https://github.com/tarantool/tarantool/issues/4148 is implemented.

**Comment by ImeevMA** [comment]
Found a bug:
```
tarantool> CREATE TABLE t(i INT PRIMARY KEY, a SCALAR);
---
- row_count: 1
...

tarantool> INSERT INTO t VALUES (1,1), (2,2), (3,3), (4,4), (5,5);
---
- row_count: 5
...

tarantool> SELECT COUNT(i) FROM t WHERE '2' > a;
---
- metadata:
  - name: COUNT(i)
    type: integer
  rows:
  - [5]
...
```
So, we can see that SCALAR rules were used here (value of type STRING always bigger that values of type INTEGER).

But:
```
tarantool> SELECT COUNT(i) FROM t WHERE '2' > a + 0;
---
- metadata:
  - name: COUNT(i)
    type: integer
  rows:
  - [1]
...
```
Here we see that '2' was implicitly cast to INTEGER and after that compared.

**Comment by ImeevMA** [comment]
Rules for comparison for SCALAR type is not defined at the moment. It will be fixed in #5148.
